core/apple/src/SourcesApple.kt

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright 2017-2023 JetBrains s.r.o. and respective authors and developers.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the LICENCE file.
+ */
+
+package kotlinx.io
+
+import kotlinx.cinterop.*
+import platform.Foundation.*
+import platform.darwin.NSInteger
+import platform.darwin.NSUInteger
+import platform.darwin.NSUIntegerVar
+import platform.posix.uint8_tVar
+import kotlin.native.ref.WeakReference
+
+/**
+ * Returns an input stream that reads from this source. Closing the stream will also close this source.
+ *
+ * @sample kotlinx.io.samples.KotlinxIoSamplesApple.asStream
+ */
+public fun Source.asNSInputStream(): NSInputStream = SourceNSInputStream(this)
+
+@OptIn(InternalIoApi::class, UnsafeNumber::class)
+private class SourceNSInputStream(
+    private val source: Source
+) : NSInputStream(NSData()), NSStreamDelegateProtocol {
+
+    private val isClosed: () -> Boolean = when (source) {
+        is RealSource -> source::closed
+        is Buffer -> {
+            { false }
+        }
+    }
+
+    private var status = NSStreamStatusNotOpen
+    private var error: NSError? = null
+        set(value) {
+            status = NSStreamStatusError
+            field = value
+            postEvent(NSStreamEventErrorOccurred)
+            source.close()
+        }
+
+    override fun streamStatus() = if (status != NSStreamStatusError && isClosed()) NSStreamStatusClosed else status
+
+    override fun streamError() = error
+
+    override fun open() {
+        if (status == NSStreamStatusNotOpen) {
+            status = NSStreamStatusOpening
+            status = NSStreamStatusOpen
+            postEvent(NSStreamEventOpenCompleted)
+            checkBytes()
+        }
+    }
+
+    override fun close() {
+        if (status == NSStreamStatusError || status == NSStreamStatusNotOpen) return
+        status = NSStreamStatusClosed
+        runLoop = null
+        runLoopModes = listOf()
+        source.close()
+    }
+
+    override fun read(buffer: CPointer<uint8_tVar>?, maxLength: NSUInteger): NSInteger {
+        if (streamStatus != NSStreamStatusOpen && streamStatus != NSStreamStatusAtEnd || buffer == null) return -1
```

Comment:
Sorry, I didn't notice it earlier: if we're in the `NSStreamStatusAtEnd` state, does it make sense to exit and abstain for calling `exhausted`?
The only way to transition into the at-end state is by exhausting the source and once it exhausted it's always exhausted (module Buffer's behavior).

Reply:
Handling the behavior of `Buffer` or potentially another `Source` that behaved similarly is why I treated `NSStreamStatusAtEnd` as not necessarily a terminal state, like `NSStreamStatusClosed` and `NSStreamStatusError`. I figure calling `exhausted` again shouldn't be problematic. The behavior of returning 0 and leaving the status in `NSStreamStatusAtEnd` will be the same, it'll just attempt the read again first. And if a user is continuing to call `read` after reaching this state, it's probably intentional. [Apple's docs](https://developer.apple.com/documentation/foundation/nsstreamstatus/nsstreamstatusatend) aren't clear about whether `NSStreamStatusAtEnd` can be temporary or should be considered terminal.